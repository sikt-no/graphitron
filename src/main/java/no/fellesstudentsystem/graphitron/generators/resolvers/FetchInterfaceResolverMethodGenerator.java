package no.fellesstudentsystem.graphitron.generators.resolvers;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import no.fellesstudentsystem.graphitron.definitions.fields.ObjectField;
import no.fellesstudentsystem.graphitron.definitions.objects.AbstractObjectDefinition;
import no.fellesstudentsystem.graphitron.definitions.objects.InterfaceDefinition;
import no.fellesstudentsystem.graphitron.definitions.objects.ObjectDefinition;
import no.fellesstudentsystem.graphitron.generators.abstractions.ResolverMethodGenerator;
import no.fellesstudentsystem.graphitron.generators.dependencies.Dependency;
import no.fellesstudentsystem.graphitron.generators.dependencies.QueryDependency;
import no.fellesstudentsystem.graphitron.schema.ProcessedSchema;
import no.fellesstudentsystem.graphitron.generators.abstractions.DBClassGenerator;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static no.fellesstudentsystem.graphitron.generators.db.FetchDBClassGenerator.SAVE_DIRECTORY_NAME;
import static no.fellesstudentsystem.graphitron.mappings.JavaPoetClassName.*;
import static org.apache.commons.lang3.StringUtils.capitalize;
import static org.apache.commons.lang3.StringUtils.uncapitalize;

/**
 * Generates resolvers for queries returning an interface. E.g. the node resolver.
 */
public class FetchInterfaceResolverMethodGenerator extends ResolverMethodGenerator<ObjectField> {
    private final ObjectDefinition localObject;

    public FetchInterfaceResolverMethodGenerator(ObjectDefinition localObject, ProcessedSchema processedSchema) {
        super(localObject, processedSchema);
        this.localObject = localObject;
    }

    @Override
    public MethodSpec generate(ObjectField target) {
        InterfaceDefinition interfaceDefinition = processedSchema.getInterface(target);
        TypeName returnClassName = interfaceDefinition.getGraphClassName();

        var spec = getDefaultSpecBuilder(target.getName(), returnClassName);

        if (!localObject.isRoot()) {
            spec.addParameter(localObject.getGraphClassName(), uncapitalize(localObject.getName()));
        }

        var inputField = target.getInputFields().get(0);
        String inputFieldName = inputField.getName();
        spec
                .addParameter(inputIterableWrap(inputField), inputFieldName)
                .addParameter(DATA_FETCHING_ENVIRONMENT.className, ENV_NAME)
                .addStatement("String tablePartOfId = $T.getTablePartOf($N)", FIELD_HELPERS.className, inputFieldName)
                .addCode("\n");

        processedSchema
                .getObjects()
                .values()
                .stream()
                .filter(it -> it.implementsInterface(interfaceDefinition.getName()))
                .sorted(Comparator.comparing(AbstractObjectDefinition::getName))
                .forEach(implementation -> {
                    var queryLocation = implementation.getName() + DBClassGenerator.FILE_NAME_SUFFIX;
                    var map = Map.ofEntries(
                            Map.entry("env", ENV_NAME),
                            Map.entry("string", STRING.className),
                            Map.entry("returnType", returnClassName),
                            Map.entry("dataloader", DATA_LOADER_FACTORY.className),
                            Map.entry("batchLoader", MAPPED_BATCH_LOADER_WITH_CONTEXT.className),
                            Map.entry("implementationClass", implementation.getGraphClassName()),
                            Map.entry("future", COMPLETABLE_FUTURE.className),
                            Map.entry("queryInstanceField", uncapitalize(queryLocation)),
                            Map.entry("implName", implementation.getName()),
                            Map.entry("inputFieldName", inputFieldName),
                            Map.entry("inputFieldNameCap", capitalize(inputFieldName)),
                            Map.entry("selectionSets", SELECTION_SET.className),
                            Map.entry("environmentUtils", ENVIRONMENT_UTILS.className),
                            Map.entry("referenceFieldName", capitalize(target.getName())),
                            Map.entry("context", Dependency.CONTEXT_NAME)
                    );
                    dependencySet.add(new QueryDependency(queryLocation, SAVE_DIRECTORY_NAME));

                    spec.beginControlFlow("if (tablePartOfId.equals($T.$N.getViewId().toString()))", TABLES.className, implementation.getTable().getMappingName());
                    spec.addStatement(CodeBlock.builder().addNamed(
                                    "return $env:N.getDataLoaderRegistry().<$string:T, $returnType:T>computeIfAbsent(tablePartOfId, name ->$W" +
                                            "$dataloader:T.newMappedDataLoader(($batchLoader:T<$string:T, $implementationClass:T>) (keys, loaderEnvironment) ->$>$W" +
                                            "$future:T.completedFuture($queryInstanceField:N.load$implName:NBy$inputFieldNameCap:NsAs$referenceFieldName:N" +
                                            "($context:N, keys, new $selectionSets:T($environmentUtils:T.getSelectionSetsFromEnvironment(loaderEnvironment))))))" +
                                            "$<$W.load($inputFieldName:N, $env:N)", map)
                            .build());
                    spec.endControlFlow();
                });

        spec.addStatement("throw new $T(\"could not find dataloader for $N with prefix \" + tablePartOfId)", ILLEGAL_ARGUMENT_EXCEPTION.className, inputFieldName);

        return spec.build();
    }

    @Override
    public List<MethodSpec> generateAll() {
        return localObject
                .getFields()
                .stream()
                .filter(ObjectField::isGenerated)
                .filter(it -> processedSchema.isInterface(it.getTypeName()))
                .map(this::generate)
                .collect(Collectors.toList());
    }

    @Override
    public boolean generatesAll() {
        return localObject
                .getFields()
                .stream()
                .filter(it -> processedSchema.isInterface(it.getTypeName()))
                .allMatch(ObjectField::isGenerated);
    }
}
