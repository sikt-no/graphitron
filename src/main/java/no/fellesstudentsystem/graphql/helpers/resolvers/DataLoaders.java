package no.fellesstudentsystem.graphql.helpers.resolvers;

import graphql.ErrorType;
import graphql.GraphqlErrorBuilder;
import graphql.relay.DefaultConnectionCursor;
import graphql.relay.DefaultEdge;
import graphql.relay.DefaultPageInfo;
import graphql.relay.Edge;
import graphql.schema.DataFetchingEnvironment;
import no.fellesstudentsystem.graphql.exception.ValidationViolationGraphQLException;
import no.fellesstudentsystem.graphql.helpers.EnvironmentUtils;
import no.fellesstudentsystem.graphql.helpers.functions.DBCount;
import no.fellesstudentsystem.graphql.helpers.functions.DBQuery;
import no.fellesstudentsystem.graphql.helpers.functions.DBQueryIterable;
import no.fellesstudentsystem.graphql.helpers.functions.DBQueryRoot;
import no.fellesstudentsystem.graphql.helpers.selection.ConnectionSelectionSet;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import no.fellesstudentsystem.graphql.relay.ConnectionImpl;
import no.fellesstudentsystem.graphql.relay.ExtendedConnection;
import org.dataloader.DataLoader;
import org.dataloader.DataLoaderFactory;
import org.dataloader.MappedBatchLoaderWithContext;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import java.util.stream.Collectors;

public class DataLoaders {
    /**
     * Create a paginated dataloader for a resolver. Typically used for pagination on non-root queries.
     * @param env Environment for the resolver.
     * @param resolveName Name of the resolver.
     * @param pageSize Size of the pages for pagination.
     * @param maxNodes Limit on how many elements may be fetched at once.
     * @param dbFunction Function to call to retrieve the query data.
     * @param countFunction Function to call to retrieve the total count of elements that could be potentially retrieved.
     * @param idFunction Function that extracts an ID from the fetched type.
     * @return A paginated dataloader.
     * @param <T> Type that the resolver fetches.
     */
    public static <T> DataLoader<String, ExtendedConnection<T>> getDataLoader(DataFetchingEnvironment env, String resolveName, int pageSize, int maxNodes, DBQueryIterable<String, T> dbFunction, DBCount<String> countFunction, Function<T, String> idFunction) {
        return env.getDataLoaderRegistry().computeIfAbsent(resolveName, name ->
                DataLoaderFactory.newMappedDataLoader((MappedBatchLoaderWithContext<String, ExtendedConnection<T>>) (keys, batchEnvLoader) ->
                        getMappedDataLoader(keys, new ConnectionSelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(batchEnvLoader)), maxNodes, pageSize, dbFunction, countFunction, idFunction)
                )
        );
    }

    /**
     * Create a dataloader for a resolver.
     * @param env Environment for the resolver.
     * @param resolveName Name of the resolver.
     * @param dbFunction Function to call to retrieve the query data.
     * @return A basic dataloader without pagination.
     * @param <T> Type that the resolver fetches.
     */
    @NotNull
    public static <T> DataLoader<String, T> getDataLoader(DataFetchingEnvironment env, String resolveName, DBQuery<String, T> dbFunction) {
        return env.getDataLoaderRegistry().computeIfAbsent(resolveName, name ->
                DataLoaderFactory.newMappedDataLoader((MappedBatchLoaderWithContext<String, T>) (keys, batchEnvLoader) ->
                        getMappedDataLoader(keys, new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(batchEnvLoader)), dbFunction)
                )
        );
    }

    /**
     * Load the data for a resolver. The result is paginated.
     * @param env Environment for the resolver.
     * @param pageSize Size of the pages for pagination.
     * @param maxNodes Limit on how many elements may be fetched at once.
     * @param dbFunction Function to call to retrieve the query data.
     * @param countFunction Function to call to retrieve the total count of elements that could be potentially retrieved.
     * @param idFunction Function that extracts an ID from the fetched type.
     * @return A paginated resolver result.
     * @param <T> Type that the resolver fetches.
     */
    @NotNull
    public static <T> CompletableFuture<ExtendedConnection<T>> loadData(DataFetchingEnvironment env, int pageSize, int maxNodes, DBQueryRoot<T> dbFunction, DBCount<String> countFunction, Function<T, String> idFunction) {
        var selectionSet = new ConnectionSelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(env));
        var dbResult = dbFunction.callDBMethod(selectionSet);
        var totalCount = countFunction.callDBMethod(Set.of(), selectionSet);
        return getPaginatedConnection(dbResult, pageSize, totalCount, maxNodes, idFunction);
    }

    /**
     * Load the data for a resolver in a lookup format.
     * @param env Environment for the resolver.
     * @param keys The lookup keys that are necessary for match rows in the result with the original resolver input.
     * @param dbFunction Function to call to retrieve the query data.
     * @return A resolver result without pagination.
     * @param <K> The key type that this lookup uses.
     * @param <V> Type that the resolver fetches.
     */
    @NotNull
    public static <K, V> CompletableFuture<List<V>> loadDataAsLookup(DataFetchingEnvironment env, List<List<K>> keys, DBQuery<String, V> dbFunction) {
        if (keys.isEmpty()) {
            return CompletableFuture.completedFuture(List.of());
        }

        var mergedKeys = mergeKeys(keys, env);

        var selectionSet = new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(env));
        var dbResult = dbFunction.callDBMethod(new HashSet<>(mergedKeys), selectionSet);
        var orderedResult = mergedKeys.stream().map(dbResult::get).collect(Collectors.toList());
        return CompletableFuture.completedFuture(orderedResult);
    }

    /**
     * Load the data for an interface resolver.
     * @param env Environment for the resolver.
     * @param table The source from which the data should be loaded.
     * @param keyToLoad The specific key that should be loaded.
     * @param dbFunction Function to call to retrieve the query data.
     * @return A resolver result without pagination.
     * @param <U> Type that the resolver fetches.
     */
    public static <T, U extends T> CompletableFuture<T> loadInterfaceData(DataFetchingEnvironment env, String table, String keyToLoad, DBQuery<String, U> dbFunction) {
        return env
                .getDataLoaderRegistry()
                .<String, T>computeIfAbsent(table, name ->
                        DataLoaderFactory.newMappedDataLoader((MappedBatchLoaderWithContext<String, U>) (ids, loaderEnvironment) ->
                                CompletableFuture.completedFuture(dbFunction.callDBMethod(ids, new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(loaderEnvironment))))
                        )
                )
                .load(keyToLoad, env);
    }

    public static <K> List<String> mergeKeys(List<List<K>> keys, DataFetchingEnvironment env) {
        if (keys.isEmpty()) {
            return List.of();
        }

        // This assumes all keys are correlated.
        var nKeys = keys.get(0).size();
        if (keys.stream().map(List::size).anyMatch(it -> it != nKeys)) {
            throw new ValidationViolationGraphQLException(
                    List.of(
                            GraphqlErrorBuilder
                                    .newError(env)
                                    .path(env.getExecutionStepInfo().getPath())
                                    .message("Keys sets have differing lengths. For this type of query, each key field is required to be an array of equal length.")
                                    .errorType(ErrorType.ValidationError)
                                    .build()
                    )
            );
        }

        var mergedKeys = new ArrayList<>(Collections.nCopies(nKeys, ""));
        for (var keyList : keys) {
            for (int i = 0; i < nKeys; i++) {
                var currentKey = mergedKeys.get(i);
                mergedKeys.set(i, (currentKey.isEmpty() ? "" : currentKey + ",") + keyList.get(i));
            }
        }

        return mergedKeys;
    }

    public static <T> CompletableFuture<T> load(DataLoader<String, T> loader, String id, DataFetchingEnvironment env) {
        return loader.load(env.getExecutionStepInfo().getPath().toString() + "||" + id, env);
    }

    public static <T> CompletableFuture<List<T>> loadNonNullable(DataLoader<String, List<T>> loader, String id, DataFetchingEnvironment env) {
        return loader.load(env.getExecutionStepInfo().getPath().toString() + "||" + id, env).thenApply(data -> Optional.ofNullable(data).orElse(List.of()));
    }

    @NotNull
    private static <T> CompletableFuture<Map<String, ExtendedConnection<T>>> getMappedDataLoader(Set<String> keys, SelectionSet selectionSet, int maxNodes, int pageSize, DBQueryIterable<String, T> dbFunction, DBCount<String> countFunction, Function<T, String> idFunction) {
        if (keys.isEmpty()) {
            return CompletableFuture.completedFuture(Map.of());
        }

        var keyToId = getKeyToId(keys);
        var idSet = new HashSet<>(keyToId.values());
        return DataLoaders.getPaginatedConnection(
                resultAsMapForConnection(keyToId, dbFunction.callDBMethod(idSet, selectionSet)),
                pageSize,
                countFunction.callDBMethod(idSet, selectionSet),
                maxNodes,
                idFunction
        );
    }

    @NotNull
    private static <T> CompletableFuture<Map<String, T>> getMappedDataLoader(Set<String> keys, SelectionSet selectionSet, DBQuery<String, T> dbFunction) {
        if (keys.isEmpty()) {
            return CompletableFuture.completedFuture(Map.of());
        }

        var keyToId = getKeyToId(keys);
        return CompletableFuture.completedFuture(resultAsMap(keyToId, dbFunction.callDBMethod(new HashSet<>(keyToId.values()), selectionSet)));
    }

    @NotNull
    private static <T> Map<String, T> resultAsMap(Map<String, String> keyToId, Map<String, T> dbResult) {
        return keyToId
                .entrySet()
                .stream()
                .filter(it -> it.getValue() != null)
                .filter(it -> dbResult.get(it.getValue()) != null)
                .collect(Collectors.toMap(Map.Entry::getKey, it -> dbResult.get(it.getValue())));
    }

    @NotNull
    private static <T> Map<String, T> resultAsMapForConnection(Map<String, String> keyToId, Map<String, T> dbResult) {
        return keyToId
                .entrySet()
                .stream()
                .filter(it -> it.getValue() != null)
                .filter(it -> dbResult.get(it.getValue()) != null)
                .collect(
                        Collectors.toMap(
                                Map.Entry::getKey,
                                it -> {
                                    var dbValue = dbResult.get(it.getValue());
                                    return dbValue instanceof List<?> ? (T)((List<?>)dbValue).stream().filter(Objects::nonNull).collect(Collectors.toList()) : dbValue;
                                }
                        )
                );
    }

    @NotNull
    private static Map<String, String> getKeyToId(Set<String> keys) {
        return keys.stream().collect(Collectors.toMap(s -> s, s -> s.substring(s.lastIndexOf("||") + 2)));
    }

    @NotNull
    public static <T> CompletableFuture<ExtendedConnection<T>> getPaginatedConnection(List<T> dbResult, int pageSize, Integer totalCount, int maxNodes, Function<T, String> idFunction) {
        return CompletableFuture.completedFuture(createPagedResult(dbResult, pageSize, totalCount != null ? Math.min(maxNodes, totalCount) : null, idFunction));
    }

    @NotNull
    public static <T> CompletableFuture<Map<String, ExtendedConnection<T>>> getPaginatedConnection(Map<String, List<T>> dbResult, int pageSize, Integer totalCount, int maxNodes, Function<T, String> idFunction) {
        var pagedResult = dbResult.entrySet().stream().map(resultEntry -> {
            var pagedResultEntry = createPagedResult(resultEntry.getValue(), pageSize, totalCount != null ? Math.min(maxNodes, totalCount) : null, idFunction);
            return new AbstractMap.SimpleEntry<String, ExtendedConnection<T>>(resultEntry.getKey(), pagedResultEntry);
        }).collect(Collectors.toMap(AbstractMap.SimpleEntry::getKey, AbstractMap.SimpleEntry::getValue));
        return CompletableFuture.completedFuture(pagedResult);
    }

    private static <T> ConnectionImpl<T> createPagedResult(List<T> dbResult, int pageSize, Integer totalCount, Function<T, String> idFunction) {
        var items = dbResult.subList(0, Math.min(dbResult.size(), pageSize));
        DefaultPageInfo pageInfo;
        if (items.isEmpty()) {
            pageInfo = new DefaultPageInfo(null, null, false, dbResult.size() > pageSize);
        } else {
            var itemStart = items.get(0);
            var itemEnd = items.get(items.size() - 1);
            pageInfo = new DefaultPageInfo(
                    new DefaultConnectionCursor(itemStart != null ? idFunction.apply(itemStart) : null),
                    new DefaultConnectionCursor(itemEnd != null ? idFunction.apply(itemEnd) : null),
                    false,
                    dbResult.size() > pageSize
            );
        }

        List<Edge<T>> edges = items
                .stream()
                .map(item -> new DefaultEdge<>(item, new DefaultConnectionCursor(item != null ? idFunction.apply(item) : null)))
                .collect(Collectors.toList());

        return ConnectionImpl
                .<T>builder()
                .setPageInfo(pageInfo)
                .setNodes(items)
                .setEdges(edges)
                .setTotalCount(totalCount)
                .build();
    }
}
