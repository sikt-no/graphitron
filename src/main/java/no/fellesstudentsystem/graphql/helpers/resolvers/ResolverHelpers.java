package no.fellesstudentsystem.graphql.helpers.resolvers;

import graphql.schema.DataFetchingEnvironment;
import no.fellesstudentsystem.graphql.helpers.EnvironmentUtils;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import org.jetbrains.annotations.NotNull;
import org.jooq.DSLContext;

import java.util.Optional;

public class ResolverHelpers {
    public static int getPageSize(Integer first, int max, int defaultMax) {
        return Optional.ofNullable(first).map(it -> Math.min(max, it)).orElse(defaultMax);
    }

    public static DSLContext selectContext(DataFetchingEnvironment env, DSLContext defaultContext) {
        return env.getLocalContext() == null ? defaultContext : (DSLContext) env.getLocalContext();
    }

    @NotNull
    public static SelectionSet getSelectionSet(DataFetchingEnvironment env) {
        return new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(env));
    }
}
