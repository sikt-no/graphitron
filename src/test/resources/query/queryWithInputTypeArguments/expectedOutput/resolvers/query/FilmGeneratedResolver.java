package fake.code.generated.resolvers.query;

import fake.code.generated.queries.query.FilmDBQueries;
import fake.graphql.example.api.FilmResolver;
import fake.graphql.example.model.Film;
import fake.graphql.example.model.FilmCategory;
import fake.graphql.example.model.OriginalCategoryInput;
import fake.graphql.example.model.OriginalCategoryInputNested;
import fake.graphql.example.model.OriginalCategoryInputWithOneField;
import graphql.schema.DataFetchingEnvironment;
import java.lang.Exception;
import java.lang.Override;
import java.lang.String;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import javax.inject.Inject;
import no.fellesstudentsystem.graphql.helpers.resolvers.DataLoaders;
import no.fellesstudentsystem.graphql.helpers.resolvers.ResolverHelpers;
import org.dataloader.DataLoader;
import org.jooq.DSLContext;

public class FilmGeneratedResolver implements FilmResolver {
    @Inject
    DSLContext ctx;

    @Inject
    private FilmDBQueries filmDBQueries;

    @Override
    public CompletableFuture<List<FilmCategory>> categories(Film film,
            OriginalCategoryInput categoryIn, DataFetchingEnvironment env) throws Exception {
        var ctx = ResolverHelpers.selectContext(env, this.ctx);
        DataLoader<String, List<FilmCategory>> loader = DataLoaders.getDataLoader(env, "categoriesForFilm", (ids, selectionSet) -> filmDBQueries.categoriesForFilm(ctx, ids, categoryIn, selectionSet));
        return DataLoaders.loadNonNullable(loader, film.getId(), env);
    }

    @Override
    public CompletableFuture<List<FilmCategory>> categoriesForInputList(Film film,
            List<OriginalCategoryInput> categoryInList, DataFetchingEnvironment env) throws
            Exception {
        var ctx = ResolverHelpers.selectContext(env, this.ctx);
        DataLoader<String, List<FilmCategory>> loader = DataLoaders.getDataLoader(env, "categoriesForInputListForFilm", (ids, selectionSet) -> filmDBQueries.categoriesForInputListForFilm(ctx, ids, categoryInList, selectionSet));
        return DataLoaders.loadNonNullable(loader, film.getId(), env);
    }

    @Override
    public CompletableFuture<List<FilmCategory>> categoriesForMixOfListAndSingleInput(Film film,
            OriginalCategoryInput categoryIn, List<OriginalCategoryInput> categoryInList,
            DataFetchingEnvironment env) throws Exception {
        var ctx = ResolverHelpers.selectContext(env, this.ctx);
        DataLoader<String, List<FilmCategory>> loader = DataLoaders.getDataLoader(env, "categoriesForMixOfListAndSingleInputForFilm", (ids, selectionSet) -> filmDBQueries.categoriesForMixOfListAndSingleInputForFilm(ctx, ids, categoryIn, categoryInList, selectionSet));
        return DataLoaders.loadNonNullable(loader, film.getId(), env);
    }

    @Override
    public CompletableFuture<List<FilmCategory>> categoriesForInputWithOneFieldList(Film film,
            List<OriginalCategoryInputWithOneField> categoryInListOneField,
            DataFetchingEnvironment env) throws Exception {
        var ctx = ResolverHelpers.selectContext(env, this.ctx);
        DataLoader<String, List<FilmCategory>> loader = DataLoaders.getDataLoader(env, "categoriesForInputWithOneFieldListForFilm", (ids, selectionSet) -> filmDBQueries.categoriesForInputWithOneFieldListForFilm(ctx, ids, categoryInListOneField, selectionSet));
        return DataLoaders.loadNonNullable(loader, film.getId(), env);
    }

    @Override
    public CompletableFuture<List<FilmCategory>> categoriesForInputWithNestedFieldList(Film film,
            List<OriginalCategoryInputNested> categoryInListNestedField,
            DataFetchingEnvironment env) throws Exception {
        var ctx = ResolverHelpers.selectContext(env, this.ctx);
        DataLoader<String, List<FilmCategory>> loader = DataLoaders.getDataLoader(env, "categoriesForInputWithNestedFieldListForFilm", (ids, selectionSet) -> filmDBQueries.categoriesForInputWithNestedFieldListForFilm(ctx, ids, categoryInListNestedField, selectionSet));
        return DataLoaders.loadNonNullable(loader, film.getId(), env);
    }
}