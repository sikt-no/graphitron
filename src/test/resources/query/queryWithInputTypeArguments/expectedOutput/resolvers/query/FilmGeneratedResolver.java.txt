package fake.code.generated.resolvers.query;

import fake.code.generated.queries.query.FilmDBQueries;
import fake.graphql.example.package.api.FilmResolver;
import fake.graphql.example.package.model.Film;
import fake.graphql.example.package.model.FilmCategory;
import fake.graphql.example.package.model.OriginalCategoryInput;
import fake.graphql.example.package.model.OriginalCategoryInputNested;
import fake.graphql.example.package.model.OriginalCategoryInputWithOneField;
import graphql.schema.DataFetchingEnvironment;
import java.lang.Exception;
import java.lang.Override;
import java.lang.String;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;
import javax.inject.Inject;
import no.fellesstudentsystem.graphql.helpers.EnvironmentUtils;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import org.dataloader.DataLoader;
import org.dataloader.DataLoaderFactory;
import org.dataloader.MappedBatchLoaderWithContext;
import org.jooq.DSLContext;

public class FilmGeneratedResolver implements FilmResolver {
    @Inject
    DSLContext ctx;

    @Inject
    private FilmDBQueries filmDBQueries;

    @Override
    public CompletableFuture<List<FilmCategory>> categories(Film film,
            OriginalCategoryInput categoryIn, DataFetchingEnvironment env) throws Exception {
        var ctx = env.getLocalContext() == null ? this.ctx : (DSLContext) env.getLocalContext();
        DataLoader<String, List<FilmCategory>> loader = env.getDataLoaderRegistry().computeIfAbsent("categoriesForFilm", name -> {
            var batchLoader = (MappedBatchLoaderWithContext<String, List<FilmCategory>>) (keys, batchEnvLoader) -> {
                var keyToId = keys.stream().collect(
                        Collectors.toMap(s -> s, s -> s.substring(s.lastIndexOf("||") + 2)));
                var idSet = new HashSet<>(keyToId.values());
                var selectionSet = new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(batchEnvLoader));
                var dbResult = filmDBQueries.categoriesForFilm(ctx, idSet, categoryIn, selectionSet);
                var mapResult = keyToId.entrySet().stream()
                        .filter(it -> dbResult.get(it.getValue()) != null)
                        .collect(Collectors.toMap(Map.Entry::getKey, it -> dbResult.get(it.getValue())));
                return CompletableFuture.completedFuture(mapResult);
            } ;
            return DataLoaderFactory.newMappedDataLoader(batchLoader);
        } );
        return loader.load(env.getExecutionStepInfo().getPath().toString() + "||" + film.getId(), env).thenApply(data -> Optional.ofNullable(data).orElse(List.of()));
    }

    @Override
    public CompletableFuture<List<FilmCategory>> categoriesForInputList(Film film,
            List<OriginalCategoryInput> categoryInList, DataFetchingEnvironment env) throws
            Exception {
        var ctx = env.getLocalContext() == null ? this.ctx : (DSLContext) env.getLocalContext();
        DataLoader<String, List<FilmCategory>> loader = env.getDataLoaderRegistry().computeIfAbsent("categoriesForInputListForFilm", name -> {
            var batchLoader = (MappedBatchLoaderWithContext<String, List<FilmCategory>>) (keys, batchEnvLoader) -> {
                var keyToId = keys.stream().collect(
                        Collectors.toMap(s -> s, s -> s.substring(s.lastIndexOf("||") + 2)));
                var idSet = new HashSet<>(keyToId.values());
                var selectionSet = new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(batchEnvLoader));
                var dbResult = filmDBQueries.categoriesForInputListForFilm(ctx, idSet, categoryInList, selectionSet);
                var mapResult = keyToId.entrySet().stream()
                        .filter(it -> dbResult.get(it.getValue()) != null)
                        .collect(Collectors.toMap(Map.Entry::getKey, it -> dbResult.get(it.getValue())));
                return CompletableFuture.completedFuture(mapResult);
            } ;
            return DataLoaderFactory.newMappedDataLoader(batchLoader);
        } );
        return loader.load(env.getExecutionStepInfo().getPath().toString() + "||" + film.getId(), env).thenApply(data -> Optional.ofNullable(data).orElse(List.of()));
    }

    @Override
    public CompletableFuture<List<FilmCategory>> categoriesForMixOfListAndSingleInput(Film film,
            OriginalCategoryInput categoryIn, List<OriginalCategoryInput> categoryInList,
            DataFetchingEnvironment env) throws Exception {
        var ctx = env.getLocalContext() == null ? this.ctx : (DSLContext) env.getLocalContext();
        DataLoader<String, List<FilmCategory>> loader = env.getDataLoaderRegistry().computeIfAbsent("categoriesForMixOfListAndSingleInputForFilm", name -> {
            var batchLoader = (MappedBatchLoaderWithContext<String, List<FilmCategory>>) (keys, batchEnvLoader) -> {
                var keyToId = keys.stream().collect(
                        Collectors.toMap(s -> s, s -> s.substring(s.lastIndexOf("||") + 2)));
                var idSet = new HashSet<>(keyToId.values());
                var selectionSet = new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(batchEnvLoader));
                var dbResult = filmDBQueries.categoriesForMixOfListAndSingleInputForFilm(ctx, idSet, categoryIn, categoryInList, selectionSet);
                var mapResult = keyToId.entrySet().stream()
                        .filter(it -> dbResult.get(it.getValue()) != null)
                        .collect(Collectors.toMap(Map.Entry::getKey, it -> dbResult.get(it.getValue())));
                return CompletableFuture.completedFuture(mapResult);
            } ;
            return DataLoaderFactory.newMappedDataLoader(batchLoader);
        } );
        return loader.load(env.getExecutionStepInfo().getPath().toString() + "||" + film.getId(), env).thenApply(data -> Optional.ofNullable(data).orElse(List.of()));
    }

    @Override
    public CompletableFuture<List<FilmCategory>> categoriesForInputWithOneFieldList(Film film,
            List<OriginalCategoryInputWithOneField> categoryInListOneField,
            DataFetchingEnvironment env) throws Exception {
        var ctx = env.getLocalContext() == null ? this.ctx : (DSLContext) env.getLocalContext();
        DataLoader<String, List<FilmCategory>> loader = env.getDataLoaderRegistry().computeIfAbsent("categoriesForInputWithOneFieldListForFilm", name -> {
            var batchLoader = (MappedBatchLoaderWithContext<String, List<FilmCategory>>) (keys, batchEnvLoader) -> {
                var keyToId = keys.stream().collect(
                        Collectors.toMap(s -> s, s -> s.substring(s.lastIndexOf("||") + 2)));
                var idSet = new HashSet<>(keyToId.values());
                var selectionSet = new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(batchEnvLoader));
                var dbResult = filmDBQueries.categoriesForInputWithOneFieldListForFilm(ctx, idSet, categoryInListOneField, selectionSet);
                var mapResult = keyToId.entrySet().stream()
                        .filter(it -> dbResult.get(it.getValue()) != null)
                        .collect(Collectors.toMap(Map.Entry::getKey, it -> dbResult.get(it.getValue())));
                return CompletableFuture.completedFuture(mapResult);
            } ;
            return DataLoaderFactory.newMappedDataLoader(batchLoader);
        } );
        return loader.load(env.getExecutionStepInfo().getPath().toString() + "||" + film.getId(), env).thenApply(data -> Optional.ofNullable(data).orElse(List.of()));
    }

    @Override
    public CompletableFuture<List<FilmCategory>> categoriesForInputWithNestedFieldList(Film film,
            List<OriginalCategoryInputNested> categoryInListNestedField,
            DataFetchingEnvironment env) throws Exception {
        var ctx = env.getLocalContext() == null ? this.ctx : (DSLContext) env.getLocalContext();
        DataLoader<String, List<FilmCategory>> loader = env.getDataLoaderRegistry().computeIfAbsent("categoriesForInputWithNestedFieldListForFilm", name -> {
            var batchLoader = (MappedBatchLoaderWithContext<String, List<FilmCategory>>) (keys, batchEnvLoader) -> {
                var keyToId = keys.stream().collect(
                        Collectors.toMap(s -> s, s -> s.substring(s.lastIndexOf("||") + 2)));
                var idSet = new HashSet<>(keyToId.values());
                var selectionSet = new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(batchEnvLoader));
                var dbResult = filmDBQueries.categoriesForInputWithNestedFieldListForFilm(ctx, idSet, categoryInListNestedField, selectionSet);
                var mapResult = keyToId.entrySet().stream()
                        .filter(it -> dbResult.get(it.getValue()) != null)
                        .collect(Collectors.toMap(Map.Entry::getKey, it -> dbResult.get(it.getValue())));
                return CompletableFuture.completedFuture(mapResult);
            } ;
            return DataLoaderFactory.newMappedDataLoader(batchLoader);
        } );
        return loader.load(env.getExecutionStepInfo().getPath().toString() + "||" + film.getId(), env).thenApply(data -> Optional.ofNullable(data).orElse(List.of()));
    }
}